### PR TITLE
Conditionally pass -- delimiter to main

### PR DESCRIPTION
**Pull request by arezaii**
This PR adds the ability to forward the `--` delimiter from the command line
to the arguments that are passed to `main`, if and only if main is declared
with arguments and the ArgumentParser package is used by the program.

These changes were prototyped during cray/chapel-private#2450 and are
implemented here similar to the prototype.

TESTING:

After running `make cleanall` from `$CHPL_HOME`:

- [x] Can `make`
- [x] Can `make docs`
- [x] Can `make mason`
- [x] Can `make check`
- [x] Passing all tests from `util/test/paratest.chapcs`

Reviewed by @mppf 

Signed-off-by: Ahmad Rezaii ahmad.rezaii@hpe.com

### CODE REVIEW

**Review by mppf — APPROVED** [review]
I'm not sure if this should go in the release or not but I was looking at it so thought I'd leave a review.

The changes seem fine with me and are definitely an improvement.

However, I wonder if we should be using a different approach. In particular, I would expect that programs using ArgumentParsing and declaring `main` with args should not have any automatic config variable parsing at all. Instead, they would parse the arguments and then pass along any unparsed arguments to the config var processing by calling a function e.g. `processConfigVarArguments(argsArray)`. This puts the author of such programs in control of the argument parsing and enables Chapel programs that replace other tools that must have specific interpretations of command line arguments. In fact, I would argue that this should be the behavior with any `main` with args and that it doesn't matter if the argument parsing module is used.

Now, I understand that getting multilocale job launch to work with that strategy is tricky. But that seems like something we could solve. Have we discussed this anywhere?

**Comment by arezaii** [comment]
>Now, I understand that getting multilocale job launch to work with that strategy is tricky. But that seems like something we could solve. Have we discussed this anywhere?

@bradcray and I had an offline discussion about this, and tried to capture the essence in the discussion on Cray/chapel-private#2450.

The roadblock we kept coming up against (as I recall) is that the runtime needs to know how many locales to launch on before the argumentparser sees the command line arguments. So there was this chicken/egg problem where we need to do some initial setup before calling main, but then main would somehow be responsible for collecting the information that was needed for the initial setup. So we ended up in this place where we decided that `-nl` would likely always have to be handled by the runtime.

**Comment by arezaii** [comment]
>I'm not sure if this should go in the release or not but I was looking at it so thought I'd leave a review.

I was hoping to have it ready in time to merge for the release, but I would understand if it can't go now

**Comment by bradcray** [comment]
Just to reinforce @arezaii's response:  We've kicked around what it would take for Chapel's automatic command-line parsing to fire after the user code, but as Michael alludes to:

> Now, I understand that getting multilocale job launch to work with that strategy is tricky.

By doing so, we'd either lose our ability to parse configs for errors in the launcher as we do today, or would have to start running general module initialization code in order to run the user's argparse in the launcher (which I think of as being sufficiently challenging / intractable to not be a viable solution).

I think of this question as being mostly orthogonal to this PR, though, in that the argparse library and Ahmad's recent work haven't really changed the status quo (Chapel gets first shot at arguments; remaining args are passed to main); it's only changed what happens to those arguments once they're sent to main.

One change to the status quo that we're considering is to have any programs that rely on `--` have to deal with that argument in `main()`, but we decided not to do that here since it could potentially break existing programs (though I'd guess probably not many in practice).

With respect to the release, I think this is technically a "new feature", but I'd be inclined to turn a blind eye to that if it could be merged today because I think it's pretty specific to the ArgParse library which is itself a new feature (and I think we could stretch things a bit and say that it's fixing the "bug" of how `--` is or isn't handled w.r.t. the argparse library and Chapel's argument parsing that exists on `main` today.  If someone felt that I'm stretching things too far in saying that, I'm definitely open to debating it, though.  Note that I haven't reviewed the PR itself, just the concept.

FWIW, my current favorite idea for telling Chapel to "parse nothing" is to do so via the import or use of a special module, but that's part of future brainstorming work.

**Comment by gbtitus** [comment]
For what it's worth, all the runtime does with the value given in the `-nl` or equivalent `--numLocales=` command line argument is effectively to store it into the `numLocales` config const and make sure it's the same value as the number of processes the comm layer says the system launcher started.  This latter comparison basically never fails when we're using a Chapel launcher, because the Chapel launcher sends the value of its `-nl` option to both the system launcher and the `_real` command line, so it's just the same value arriving via two different paths.  I guess it has some value as a sanity check in that you'd get an error message if there were a bug somewhere in the Chapel stack or system launcher, but I can't recall ever seeing that in > 10 years on the project.  I have seen this check fail when I ran the system launcher manually and gave different process counts to it and the `_real`, however, so it has some value for that case.

An alternative design would be to throw out the `-nl` command line option and have the runtime take the system launcher's process count as authoritative, including setting `numLocales` from it. I would even be fine with making the `numLocales` config "special" in the sense of not letting it be set from the command line. This would be Very Easy ™️  to do, but not for this release.

**Comment by bradcray** [comment]
I think that parsing the numlocales / -nl flags are the easy part.  If we wanted to make it even easier, I could even imagine requiring those flags to be the first on a Chapel executable's command-line when used (and have mentioned this to Ahmad).

For me, where the pain starts is that we currently parse things like `--debug=false` in the launcher so that if you accidentally typed `--dbug=false` or `--debug=flase`, you'll get an error immediately rather than after waiting in a slurm queue, finally getting your 1024 nodes, and then having all of them realize that you had a trivial typo.  That's a pretty huge productivity win that I'm loathe to throw out without a really strong rationale and conversation about it.  And I don't see a path to giving the user's argument parser first crack at these variables without running arbitrary Chapel module initialization and main() code in the launcher, which is why I balk at that approach.

**Comment by mppf** [comment]
Right, but effectively with main-with-args and the ArgParse module, the user is saying that they want to parse the arguments. These non-config arguments aren't currently checked for errors before nodes are requested etc with the launcher.

Meaning, the direction I would go here is:
 * for programs relying on `config` variables for argument parsing, the launcher can and should continue to do some validity checking on the arguments
 * for programs relying on main-with-args and ArgParse, that is no longer possible in general, so let's stop checking the config vars in the launcher in that setting in order to give the programmer more control over the acceptable command lines

**Comment by bradcray** [comment]
> the direction I would go here is

@mppf: Do you agree that this part of the conversation is orthogonal to this PR / the current ArgParse effort (in that what we've done is no different than the status quo that Chapel already had w.r.t. greedily doing its argument parsing first and then passing the remainder to main()?).  I suppose the one way it's related is that we are changing how `--` is passed to main in this PR, but only when using the argument parsing library (the first thing that has arguably wanted it), but beyond that, I feel like you're tackling design questions that we deliberately decided we were not going to wrestle with in this release.  So I'd be more comfortable tackling them on a future-looking design issue than on this PR.

**Comment by mppf** [comment]
@bradcray - yep. I'm comfortable with this PR going in as-is for the release and for the design discussion moving to a separate design issue (which I don't think we have yet).

File: test/library/packages/ArgumentParser/UsageTest.bad.pt.shadow.good [hunk 1]
```diff
@@ -0,0 +1,13 @@
+Config Var Values:
```
Comment (cassella) [inline test/library/packages/ArgumentParser/UsageTest.bad.pt.shadow.good:1]: Hi @arezaii -- I don't think this .good file is used by any tests.  Am I missing it?  (Also, this .good file is identical to `ArgumentParserUsageTest.PtShadow.good`)

Comment (arezaii) [reply test/library/packages/ArgumentParser/UsageTest.bad.pt.shadow.good:1]: Hi @cassella, it looks like I moved the test but I didn't remove the .good file. Thanks for catching this.

**Pull request merged**